x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/queries/TextExpansionQueryBuilderTests.java

```diff
@@ -0,0 +1,121 @@
+import org.elasticsearch.xpack.core.ml.action.InferModelAction;
+import org.elasticsearch.xpack.core.ml.inference.results.SlimResults;
+import org.elasticsearch.xpack.ml.MachineLearning;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class TextExpansionQueryBuilderTests extends AbstractQueryTestCase<TextExpansionQueryBuilder> {
+
+    @Override
+    protected TextExpansionQueryBuilder doCreateTestQueryBuilder() {
+        var builder = new TextExpansionQueryBuilder(
+            randomAlphaOfLength(4),
+            randomAlphaOfLength(4),
+            randomBoolean() ? null : randomAlphaOfLength(4)
+        );
+        if (randomBoolean()) {
+            builder.boost((float) randomDoubleBetween(0.1, 10.0, true));
+        }
+        if (randomBoolean()) {
+            builder.queryName(randomAlphaOfLength(4));
+        }
+        return builder;
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return List.of(MachineLearning.class);
+    }
+
+    @Override
+    public void testMustRewrite() {
+        SearchExecutionContext context = createSearchExecutionContext();
+        TextExpansionQueryBuilder builder = new TextExpansionQueryBuilder("foo", "bar", "baz");
+        IllegalStateException e = expectThrows(IllegalStateException.class, () -> builder.toQuery(context));
+        assertEquals("terms query must not be null, missing rewrite?", e.getMessage());
+    }
+
+    @Override
+    protected boolean canSimulateMethod(Method method, Object[] args) throws NoSuchMethodException {
+        return method.equals(Client.class.getMethod("execute", ActionType.class, ActionRequest.class, ActionListener.class))
+            && (args[0] instanceof InferModelAction);
+    }
+
+    @Override
+    protected Object simulateMethod(Method method, Object[] args) {
+        InferModelAction.Request request = (InferModelAction.Request) args[1];
+
+        var tokens = new ArrayList<SlimResults.WeightedToken>();
+        int numTokens = randomIntBetween(1, 20);
+        for (int i = 0; i < numTokens; i++) {
+            tokens.add(new SlimResults.WeightedToken(i, randomFloat()));
+        }
+
+        var response = InferModelAction.Response.builder()
+            .setModelId(request.getModelId())
+            .addInferenceResults(List.of(new SlimResults("foo", tokens, randomBoolean())))
+            .build();
+        @SuppressWarnings("unchecked")  // We matched the method above.
+        ActionListener<InferModelAction.Response> listener = (ActionListener<InferModelAction.Response>) args[2];
+        listener.onResponse(response);
+        return null;
+    }
+
+    @Override
+    protected void doAssertLuceneQuery(TextExpansionQueryBuilder queryBuilder, Query query, SearchExecutionContext context)
+        throws IOException {
+        fail("what should be asserted here?");
```

Comment:
I'm not sure what should be tested here. Any suggestions would are welcome

Reply:
Here is what I would check:

 - `query instanceof BooleanQuery`
 - Check `BooleanQuery#clauses(), hasSize(numTokens)`
 - Verify each clause is `SHOULD`
 - If you know it, the verify clause terms & boosts

Reply:
`query` is always an instance of `MatchNoDocsQuery`. Am I missing some setup for this test?

Reply:
yep, just commented. Take a look at RankFeatureQueryBuilderTests and its overridden methods.